### DISCUSSION TITLE
Unable to connect to unleash server on Unity Android App

### DISCUSSION THREAD

**Discussion by rahul-Inspirit** [Unleash/unleash#4605] [Q&A] [answered]
Hi team,

I'm using Unity Game Engine to develop an android app. Using [this](https://github.com/Unleash/unleash-client-dotnet) package I'm able to connect to Unleash server and fetch toggle payloads, but it works only in the Unity Editor (I'm using a Mac). 

When I build and run the apk on the mobile device, I'm unable to connect to Unleash server. There are no errors in the android logcat. I'm using Serilog package for Unleash logs. I do get connection errors in the Unity Editor if there is something wrong in Unleash settings.

In the mobile device, execution hangs at this line
`IUnleash unleash = await unleashFactory.CreateClientAsync(settings, synchronousInitialization: true);`

I'm not able to understand the cause of this issue. Please HELP!

**Comment by daveleek** [comment]
Hello @rahul-Inspirit!

Could you tell me a little bit more about your setup? Is Unleash hosted locally on the same network as your Mac?
How does the configuration get applied to Unleash settings, and have you configured a custom timeout setting on the UnleashSettings? By default it should time out after 5 seconds

**Reply by rahul-Inspirit** [reply]
Hi @daveleek!

We've self hosted Unleash. It is not on the same network as Mac. It has a different URL. 

I'm using the following code to configure settings:
```NET
var settings = new UnleashSettings()
{
    AppName = "dotnet-test",
    UnleashApi = new Uri("<your-api-url>"),
    CustomHttpHeaders = new Dictionary<string, string>()
    {
       {"Authorization","<your-api-token>" }
    }
};
var unleashFactory = new UnleashClientFactory();

IUnleash unleash = await unleashFactory.CreateClientAsync(settings, synchronousInitialization: true);

// this `unleash` has successfully fetched feature toggles and written them to its cache.
// if network errors or disk permissions prevented this from happening, the above await would have thrown an exception

var awesome = unleash.IsEnabled("SuperAwesomeFeature");
```

I haven't configured any custom timeout.
I'm trying to connect to the Unleash server from a mobile device.

**Reply by daveleek** [reply]
Thanks @rahul-Inspirit!

Ok, so a quick recap:

This code works when running locally on your Mac in Unreal Editor.

It hangs on `CreateClientAsync` after you build and deploy it on an android unit and then run it.
On the android unit when running you have evidence that it has fetched the toggles from Unleash and written it to cache.

Could you try something like adding a `ConfigureAwait(false)` to the `await unleashFactory.CreateClientAsync(settings, synchronousInitialization: true);` call? I might be completely wrong on this, but it feels a bit like an async await deadlock.

Another thing you can try is to use the plain `CreateClient` method on the factory with the same parameters, and see if you get further with that. Would be interesting to know if that also hangs

**Reply by rahul-Inspirit** [reply]
Thanks @daveleek 

Yes, the code works when running locally on my Mac in Unity (NOT Unreal) editor.
The same code doesn't work on android.

I don't have evidence that it has fetched the toggles from Unleash and written it to cache, I think it doesn't even connect to Unleash. 
I tried the `CreateClient` method. It also runs fine in the editor but gives the following exception on android:
```
2023-09-05 18:54:46.709 5780 5800 Error Unity IndexOutOfRangeException: Index was outside the bounds of the array.

2023-09-05 18:54:46.709 5780 5800 Error Unity   at Microsoft.CSharp.RuntimeBinder.Semantics.TypeParameterType.get_AssociatedSystemType () [0x00000] in <00000000000000000000000000000000>:0 

2023-09-05 18:54:46.709 5780 5800 Error Unity   at Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadClassTypeParameter (Microsoft.CSharp.RuntimeBinder.Semantics.AggregateSymbol parent, System.Type t) [0x00000] in <00000000000000000000000000000000>:0 

2023-09-05 18:54:46.709 5780 5800 Error Unity   at Microsoft.CSharp.RuntimeBinder.SymbolTable.LoadSymbolsFromType (System.Type type) [0x00000] in <00000000000000000000000000000000>:0 

2023-09-05 18:54:46.709 5780 5800 Error Unity   at Microsoft.CSharp.RuntimeBinder.SymbolTable.GetConstructedType (System.Type type, Microsoft.CSharp.RuntimeBinder.Semantics.AggregateSymbol agg) [0x00000] in <00000000000000000000000000000000>:0 

2023-09-05 18:54:46.709 5780 5800 Error Unity   at Microsoft.CSharp.RuntimeBinder.SymbolTable.AddMethodToSymbolTable (System.Reflection.MethodBase member, Microsoft.CSharp.RuntimeBinder.Semantics.AggregateSymbol callingAggregate, Microsoft.CSharp.RuntimeBinder.Semantics.MethodKindEnum kind) [0x00000] in <00000000000
```

**Reply by mvittiglio** [reply]
Hey @rahul-Inspirit, what version of Unity and Android are we working with? I'm hoping to try and replicate the behavior.

**Reply by rahul-Inspirit** [reply]
Hi @mvittiglio 
Unity version - 2021.3.23f1 (Silicon)
Android version - 12

**Reply by gardleopard** [reply, marked as answer]
Just a heads-up, we haven't had the chance to test or officially support using Unity for Android with the .NET SDK.
If you're a paying Unleash customer and you're keen on this, please give us a shout through our support. Your feedback matters a lot, and we'll look into prioritizing it based on your feedback.
Thanks for being part of the Unleash community, and we're all ears for your thoughts.
